### PR TITLE
Add exponential retry on network errors

### LINKED ISSUE #98: Retry feature
Networks suck. They fail all the time and are fragile. Having functionality for `got` to retry in case of failure would make it a lot more solid.

@vdemedes made this great module we could use a inspiration: https://github.com/vdemedes/got-retry

Question: Should it be on by default?

-

@floatdrop Thoughts?

-

Related: https://github.com/request/request/issues/1300

### PR DESCRIPTION
**Pull request by floatdrop**
Basic retry functional.

//cc @vdemedes 

Closes #98

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,11 +19,21 @@ var nodeStatusCodes = require('node-status-codes');
 var isPlainObj = require('is-plain-obj');
 var parseJson = require('parse-json');
 
+function backoff(iter) {
+	var noise = Math.random() * 1000;
+	return (1 << iter) * 1000 + noise;
+}
+
+function isNetworkError(err) {
+	return err.code === 'EAI_AGAIN' || err.code === 'ETIMEDOUT' || err.code === 'ESOCKETTIMEDOUT';
```
Comment (sindresorhus) [inline index.js]: Where did you find these?

Comment (floatdrop) [inline index.js]: `ETIMEDOUT` and `ESOCKETTIMEDOUT` is from [timed-out](https://github.com/floatdrop/timed-out), `EAI_AGAIN` from issue. Is there definitive list of those?

File: test/retry.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		res.end('who`s there?');
+	}
+});
+
+test.before('retry - setup', t => {
+	s.listen(s.port, () => t.end());
+});
+
+test('retry - timeout errors', t => {
+	got(s.url + '/knock-twice', {timeout: 1000}, (err, data) => {
```
Comment (sindresorhus) [inline test/retry.js]:
```js
got(`${s.url}/knock-twice`,
```

File: test/retry.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+		t.ifError(err);
+		t.is(data, 'who`s there?');
+		t.end();
+	});
```
Comment (sindresorhus) [inline test/retry.js]: I think we should start using the promise interface as default for new tests. It makes the test a lot simpler.

This one can be:

```js
test('retry - timeout errors', async t => {
	t.is(await got(`${s.url}/knock-twice`, {timeout: 1000}), 'who`s there?');
});
```

File: test/retry.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	s.listen(s.port, () => t.end());
+});
+
+test('retry - timeout errors', t => {
+	got(`${s.url}/knock-twice`, {timeout: 1000}, (err, data) => {
+		t.ifError(err);
+		t.is(data, 'who`s there?');
+		t.end();
+	});
+});
```
Comment (sindresorhus) [inline test/retry.js]: I think we should start using the promise interface as default for new tests. It makes the test a lot simpler.

This one can be:

```js
test('retry - timeout errors', async t => {
	t.is(await got(`${s.url}/knock-twice`, {timeout: 1000}), 'who`s there?');
});
```

File: readme.md [hunk 5]
```diff
@@ -111,6 +111,14 @@ Type: `number`
 
 Milliseconds after which the request will be aborted and an error event with `ETIMEDOUT` code will be emitted.
 
+###### retries
+
+Type: `number`  
+Default: `5`
+
+Number of retries for network errors (`ETIMEDOUT`, `EAI_AGAIN`, `ESOCKETTIMEDOUT`).
```
Comment (sindresorhus) [inline readme.md]: Do we really have to expose this option? I'd rather have it as a good default.

File: test/retry.js (hunk 3 above)
Comment (floatdrop) [inline test/retry.js]: > I think we should start using the promise interface as default for new tests. It makes the test a lot simpler.

Yup, it looks way better. I guess this would be in a separate PR.

File: readme.md (hunk 5 above)
Comment (floatdrop) [inline readme.md]: Yes. I think there must be a way to disable retries.

File: test/retry.js (hunk 3 above)
Comment (sindresorhus) [inline test/retry.js]: Yeah: https://github.com/sindresorhus/got/issues/115

Though, you can still change this one, though, as we already have it:

```js
test('retry - timeout errors', async t => {
	t.is(await got(`${s.url}/knock-twice`, {timeout: 1000}), 'who`s there?');
});
```

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: Don't know. Maybe @vdemedes does?

File: readme.md (hunk 5 above)
Comment (sindresorhus) [inline readme.md]: Ok

File: index.js (hunk 1 above)
Comment (floatdrop) [inline index.js]: On second though, we should retry on any error in `error` handler of http.request. There also `ECONNRESET` and `ECONNREFUSED` that will occur, when remote server is restarting.

File: readme.md [hunk 6]
```diff
@@ -111,6 +111,14 @@ Type: `number`
 
 Milliseconds after which the request will be aborted and an error event with `ETIMEDOUT` code will be emitted.
 
+###### retries
+
+Type: `number`  
+Default: `5`
+
+Number of request retries when error happens.
```
Comment (sindresorhus) [inline readme.md]: `error` => `network errors`?

File: index.js (hunk 1 above)
Comment (vdemedes) [inline index.js]: EAI_AGAIN and ETIMEDOUT cause the biggest pain in the ass. ECONNRESET and ECONNREFUSED also happen, but rarely. As for ESOCKETTIMEDOUT, never seen this one.

File: readme.md (hunk 5 above)
Comment (vdemedes) [inline readme.md]: Would be definitely good to have retries option.

File: readme.md (hunk 6 above)
Comment (floatdrop) [inline readme.md]: ok
